### ISSUE TITLE
Mapped generic type loses types during inference

### ISSUE THREAD

**Issue by Jamesernator** [microsoft/TypeScript#56861]
### 🔎 Search Terms

mapped tuple type, const inference

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about mapped types

### ⏯ Playground Link

https://www.typescriptlang.org/play?ssl=13&ssc=1&pln=1&pc=1#code/MYGwhgzhAEAiCmwD2ATeAnAPAFQHzQG9oBfAKFNEhmwFcAHEeBZNLZAOwgBdpaH5sATzrxo8AB5d47FDABK8MCiTsQggILp0YQZnY0AtgCMMufBKkyYzVBhz1GQkfgKlo76B27oawLknQACmkudEEbVggALkJoAG0AaWgAS3ZoAGt4QSQAM14HAWF4AF0YiLs+RyLE4vxiAEpCNw8WiHoMQPqAbmb3MjJSNEp0US8eIdt0AEYyxEm9QxN0XB6h8BHPFW5oCdYAJlmWO29UgHMV0gB6S+gASXYcjBGUaH98-nKsfWMMONqKLY8LgFT7QAC80HY8AA7u9GJ9AnFdhgpgAaHZzfbFbqkIA

### 💻 Code

```ts
class Decoder<T> { }

class TupleDecoder<const TupleType extends ReadonlyArray<number>> extends Decoder<TupleType> {
    constructor(entryDecoders: { [K in keyof TupleType]: Decoder<TupleType[K]> }) {
        super();
    }
}

declare const decoder1: Decoder<number>;
declare const decoder2: Decoder<string>;
// Inferred to TupleDecoder<number[]>
const tupleDecoder = new TupleDecoder([decoder1, decoder2]);
```

### 🙁 Actual behavior

The type of `tupleDecoder` is inferred to `TupleDecoder<number[]>`

### 🙂 Expected behavior

It should be inferred to be `TupleDecoder<readonly [number, string]>`.

### Additional information about the issue

I believe this problem is probably two seperate problems.

One is that `const ... extends ReadonlyArray<any>` isn't being considered for mapped types.

Second is that inference is not picking up the string type in the array due to the mapped type.

**Comment by jcalz** [comment]
`Decoder<T>` has no structural dependence on `T` so there's nothing useful to infer from, see [the FAQ on unused type parameters](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-type-inference-work-on-this-interface-interface-foot--).  If you fix this then you see that `decoder2`, a `Decoder<string>`, does not satisfy the `Decoder<number>` constraint.  If you fix that by making the constraint `any` instead of `number`, then you get `TupleDecoder<(string | number)[]>`. 

That gets you to "`const T extends ⋯`" not working well for mapped types, which looks like #53813, which currently I'd work around by using variadic tuple types, giving this:

```ts
class Decoder<T> {
    declare prop: T; // you need a structural dependency 
}

class TupleDecoder<const TupleType extends ReadonlyArray<any>> extends Decoder<TupleType> {
    //                                not number ----->  ^^^
    constructor(entryDecoders: [...{ [K in keyof TupleType]: Decoder<TupleType[K]> }]) {
        //                     ^^^^ <-- variadic tuple to hint you want a tuple --->^
        super();
    }
}

declare const decoder1: Decoder<number>;
declare const decoder2: Decoder<string>;
const tupleDecoder = new TupleDecoder([decoder1, decoder2]);
// const tupleDecoder: TupleDecoder<[number, string]>
```
[Playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAiCmwD2ATeAnAPAFQHzQG8AoaU6NUMdeaAB3SVoC5psBuaAek+gE8kArtAB28eCmhhoEAC7oBwGQPRgQ5eLXjC0w4L2hEAvkSKUorAbRDwEyNFmTDZFq-Gy9N0eAA8ZWlDAASvBgKEjCILwAgugqvJhgwry4+D5+2jC2qBg4ltbumvjEZFw8JeUVlVUiSDIiAgC2AEYY0AC0HR340AB6fSRkjrLyikjoABRacrxZ9hAsANoAdCsE0AsA0tAAlsLQANbw-ABmLvke8AC6LLM52HluF5uX+IaXAJSEAxXc1X+kfT60EwHWgADcqNtQttgNAlK44UhoAALXZ1fhCADuiTqUnh1naXR633KEEsGHG7zYJOMxiIFHA1GgQzqFGy6AAjDdEOzMMJGi10LhqQyqDQWeo7BgAEzcqVYYa7ADmwtM4Wc+JsPPs0AAvCJ4Jizlr5eMFmz7ByADSS9nSj7U34SzW3dAse6uV2YBb85oYG2K4RKl4mIA)

To recap, the only problem I see here is the one that looks like a duplicate of #53813.  Everything else behaves as expected.

**Comment by Andarist** [comment]
And with https://github.com/microsoft/TypeScript/pull/56555 you won't even have to "hint" that you want a tuple here - since the usage of a mapped type will already be treated as such a hint: [TS playground](https://www.typescriptlang.org/play?ts=5.4.0-dev.20231224#code/CYUwxgNghgTiAEkoGdnwCLgPahgHgBUA+eAbwCh54B9AgLngIG5yBfc8pVRgVwAcIITGBwh8lRFgB2yAC68BIAgE8+CEAA9ZIKcDQAlEFGDSIygIIwYUZXh5SA1lKwB3KUQA05Epu260wqL4BPyCKmokFFQiMrIwPGCyWDAAFDpxyoG4yAxRVPAA2gDS8ACWUvAOIMpYAGYKYaogALoMWWKEoUpNxc1ELFSsAJRkElTI-GIpQwPw7OzkoEhwkrHwS0EAjG3YuHhSPAC2AEZi-Yvg0Csxcuu7YgBMOyJ7cjDlAObnN-KyXe0weAAXngUhALgaQnuqQKG1wmw8dxej2aM3IAHp0fl4AA9AD85CAA)
